### PR TITLE
Support gauge_int as a pseudo type that can be interpreted differently.

### PR DESCRIPTION
**Pull request by bmoyles0117**
I don't necessarily view this as the actual implementation, but a gateway to discuss how to best accomplish the goal of converting types as I've done here.

### CODE REVIEW

File: src/daemon/plugin.h [hunk 1]
```diff
@@ -42,17 +42,26 @@
...
 
-#define DS_TYPE_COUNTER  0
-#define DS_TYPE_GAUGE    1
-#define DS_TYPE_DERIVE   2
-#define DS_TYPE_ABSOLUTE 3
+#define DS_TYPE_COUNTER   0
+#define DS_TYPE_GAUGE     1
+#define DS_TYPE_DERIVE    2
+#define DS_TYPE_ABSOLUTE  3
+#define DS_TYPE_GAUGE_INT 4
```
Comment (igorpeshansky) [inline src/daemon/plugin.h]: We don't want to be changing the basic `collectd` structures -- this is about the type as it's being sent into the `createTimeSeries` endpoint, so I would instead just add a metadata key to override the value type (`double`->`int`) while keeping the value kind (`gauge` vs. `cumulative`).

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: src/write_gcm.c [hunk 2]
```diff
@@ -76,6 +76,8 @@ static const char this_plugin_name[] = "write_gcm";
 // sent to the GCMv3 API instead of the Agent Translation Service.
 static const char custom_metric_key[] = "stackdriver_metric_type";
 
+static const char custom_metric_value_type[] = "stackdriver_metric_value_type";
```
Comment (igorpeshansky) [inline src/write_gcm.c]: `custom_metric_value_type_key`.

File: src/write_gcm.c [hunk 3]
```diff
@@ -1412,7 +1415,8 @@ static void wg_payload_destroy(wg_payload_t *list) {
 // us remember that it is a compile-time string constant which does not need to
 // be copied/deallocated.
 static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
-    int ds_type, value_t value, const char **dataSourceType_static) {
+    int ds_type, char *metric_value_type, value_t value,
```
Comment (igorpeshansky) [inline src/write_gcm.c]: I'd s/`metric_value_type`/`value_type`/. Also, why not make it an `enum` instead of a string (i.e., parse the string earlier)? That way you can also add a static array mapping these `enum` values (e.g., `INT64` and `DOUBLE`) into triples of `"int64value"`/ `"doubleValue"`, `"INT64"`/`"DOUBLE"`, and the corresponding format strings...

File: src/write_gcm.c [hunk 4]
```diff
@@ -1421,8 +1425,14 @@ static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
         return -1;
       }
       *dataSourceType_static = "gauge";
-      result->field_name_static = "doubleValue";
-      snprintf(buffer, sizeof(buffer), "%f", value.gauge);
+      if(metric_value_type != NULL &&
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Missing space after `if` and trailing whitespace on the line.

File: src/write_gcm.c [hunk 5]
```diff
@@ -1421,8 +1425,14 @@ static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
...
+         strcmp(metric_value_type, "INT64") == 0) {
+        result->field_name_static = "int64Value";
+        snprintf(buffer, sizeof(buffer), "%" PRIi64, (int64_t)value.gauge);
+      } else {
+        result->field_name_static = "doubleValue";
+        snprintf(buffer, sizeof(buffer), "%f", value.gauge);
+      }
       break;
     }
     case DS_TYPE_COUNTER: {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Don't we want symmetric changes for counter/derive?

File: src/write_gcm.c [hunk 6]
```diff
@@ -3236,6 +3250,22 @@ static int wg_json_CreateTimeSeries(
       }
     }
 
+    if (head->values[0].ds_type == DS_TYPE_ABSOLUTE) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Why did you move this? These checks are independent of the metadata values.

File: src/write_gcm.c [hunk 7]
```diff
@@ -3198,21 +3227,6 @@ static int wg_json_CreateTimeSeries(
...
-        && !isfinite(head->values[0].val.gauge)) {
-      DEBUG("write_gcm: plugin: %s, plugin_type: %s, metric_type: %s, "
-            "type_instance: %s skipping non-finite gauge value %lf.",
-            head->key.plugin, head->key.plugin_instance, head->key.type,
-            head->key.type_instance, head->values[0].val.gauge);
-      continue;
-    }
 
     for (int i = 0; i < head->key.num_metadata_entries; ++i) {
       wg_metadata_entry_t *entry = &head->key.metadata_entries[i];
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Don't we need to check that the value_type is valid as well, if present?

File: src/write_gcm.c [hunk 8]
```diff
@@ -3244,12 +3274,18 @@ static int wg_json_CreateTimeSeries(
...
 
+    char *metric_value_type = wg_metric_value_type_from_wg_payload_t(head);
+
     switch (head->values[0].ds_type) {
       case DS_TYPE_GAUGE:
       wg_json_string(jc, "metricKind");
       wg_json_string(jc, "GAUGE");
       wg_json_string(jc, "valueType");
-      wg_json_string(jc, "DOUBLE");
+      if (metric_value_type != NULL) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: I'm kind of on the fence about this one. We could either leave these conditionals everywhere, or embed the logic for figuring out the default type into `wg_metric_value_type_from_wg_payload`, so it defaults gauges to `"DOUBLE"` and counters/derives to `"INT64"`. Up to you.

File: src/write_gcm.c [hunk 9]
```diff
@@ -3449,9 +3486,10 @@ static void wg_json_CollectdValues(json_ctx_t *jc,
     const wg_payload_value_t *value = &element->values[i];
 
     wg_typed_value_t typed_value;
+    char *metric_value_type = wg_metric_value_type_from_wg_payload_t(element);
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Wait, we don't need to do this for collectdTimeSeries -- only for timeSeries.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: src/write_gcm.c [hunk 10]
```diff
@@ -3105,6 +3146,53 @@ static void wg_json_Metric(json_ctx_t *jc,
   wg_json_map_close(jc);
 }
 
+static int wg_metric_value_type_from_wg_payload_t(
+    const wg_payload_t *element) {
+  int metric_value_type = -1;
```
Comment (igorpeshansky) [inline src/write_gcm.c]: `UNKNOWN`, as suggested above.

File: src/write_gcm.c [hunk 11]
```diff
@@ -3105,6 +3146,53 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+      } else {
+        WARNING("write_gcm: unsupported metric value type provided: %s", 
+                entry->value.value_text);
+      }
+      break;
+    }
+  }
+
+  // Failover to sane defaults for each DS type.
+  if(metric_value_type == -1) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Missing space after `if`.

Comment (igorpeshansky) [inline src/write_gcm.c]: `UNKNOWN`...

File: src/write_gcm.c [hunk 12]
```diff
@@ -3105,6 +3146,53 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+    switch (element->values[0].ds_type) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: This switch would need to be a separate helper function that you call from `wg_json_CollectdValues` (e.g., `wg_default_metric_value_type_from_wg_payload_t`).

File: src/write_gcm.c [hunk 13]
```diff
@@ -3105,6 +3146,53 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+    }
+  }
+
+  return metric_value_type;
+}
+
+static char *wg_metric_value_type_to_string(int metric_value_type) {
+  if (metric_value_type < 0 || 
+      metric_value_type > STATIC_ARRAY_SIZE(metric_value_type_strings)) {
+    WARNING("write_gcm: invalid metric value type converted to a string: %d",
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Maybe even `ERROR` (since this would be a bug in the code).

File: src/write_gcm.c [hunk 14]
```diff
@@ -3121,11 +3209,13 @@ static void wg_json_Points(json_ctx_t *jc, const wg_payload_t *element) {
...
 
+  int metric_value_type = wg_metric_value_type_from_wg_payload_t(element);
+
   wg_typed_value_t typed_value;
   // We don't care what the name of the type is.
   const char *data_source_type_static;
-  if (wg_typed_value_create_from_value_t_inline(&typed_value,
-            value->ds_type, value->val, &data_source_type_static) != 0) {
+  if (wg_typed_value_create_from_value_t_inline(&typed_value, value->ds_type, 
+      metric_value_type, value->val, &data_source_type_static) != 0) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: In reformatting this, you lost the indent. You need 2 spaces beyond the start of the function name (see, e.g., line 3219 below)...

File: src/write_gcm.c [hunk 15]
```diff
@@ -3244,26 +3335,30 @@ static int wg_json_CreateTimeSeries(
     wg_json_string(jc, "metric");
     wg_json_Metric(jc, head);
 
+    char *metric_value_type_string = wg_metric_value_type_to_string(
```
Comment (igorpeshansky) [inline src/write_gcm.c]: How do we handle the `NULL` that you return in case of an error?

File: src/write_gcm.c [hunk 16]
```diff
@@ -3244,26 +3335,30 @@ static int wg_json_CreateTimeSeries(
...
+        wg_json_string(jc, "CUMULATIVE");
+        wg_json_string(jc, "valueType");
+        wg_json_string(jc, metric_value_type_string);
+        break;
     }
 
     wg_json_string(jc, "points");
     wg_json_Points(jc, head);
 
+    sfree(metric_value_type_string);
```
Comment (igorpeshansky) [inline src/write_gcm.c]: This is pointing at static strings, and shouldn't be freed.

File: src/write_gcm.c [hunk 17]
```diff
@@ -3449,9 +3544,10 @@ static void wg_json_CollectdValues(json_ctx_t *jc,
     const wg_payload_value_t *value = &element->values[i];
 
     wg_typed_value_t typed_value;
+    int metric_value_type = wg_metric_value_type_from_wg_payload_t(element);
     const char *data_source_type_static;
-    if (wg_typed_value_create_from_value_t_inline(&typed_value,
-        value->ds_type, value->val, &data_source_type_static) != 0) {
+    if (wg_typed_value_create_from_value_t_inline(&typed_value, value->ds_type, 
+        metric_value_type, value->val, &data_source_type_static) != 0) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Same here: missing indent.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: src/write_gcm.c (hunk 2 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c [hunk 18]
```diff
@@ -105,6 +105,17 @@ static const char metadata_agent_default_url[] =
 // bubble it back up to the exit status of collectd.
 static _Bool wg_some_error_occurred_g = 0;
 
+// Supported metric value types that can be sent to GCM.
+#define METRIC_VALUE_TYPE_BOOL   0
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c [hunk 19]
```diff
@@ -1441,14 +1454,32 @@ static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
...
-      result->field_name_static = "int64Value";
-      snprintf(buffer, sizeof(buffer), "%" PRIi64, (int64_t)value.counter);
+      switch (metric_value_type) {
+        case METRIC_VALUE_TYPE_INT64:
+          result->field_name_static = "int64Value";
+          snprintf(buffer, sizeof(buffer), "%" PRIi64, (int64_t)value.counter);
+          break;
+        case METRIC_VALUE_TYPE_DOUBLE:
+          result->field_name_static = "doubleValue";
+          snprintf(buffer, sizeof(buffer), "%f", (float)value.counter);
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c [hunk 20]
```diff
@@ -1441,14 +1454,32 @@ static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
...
       break;
     }
     case DS_TYPE_DERIVE: {
       *dataSourceType_static = "derive";
       result->field_name_static = "int64Value";
       snprintf(buffer, sizeof(buffer), "%" PRIi64, value.derive);
+      switch (metric_value_type) {
+        case METRIC_VALUE_TYPE_INT64:
+          result->field_name_static = "int64Value";
+          snprintf(buffer, sizeof(buffer), "%" PRIi64, (int64_t)value.derive);
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c [hunk 21]
```diff
@@ -1441,14 +1454,32 @@ static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
...
+          break;
+        case METRIC_VALUE_TYPE_DOUBLE:
+          result->field_name_static = "doubleValue";
+          snprintf(buffer, sizeof(buffer), "%f", (float)value.derive);
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c (hunk 10 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c [hunk 22]
```diff
@@ -3115,21 +3146,53 @@ static void wg_json_Metric(json_ctx_t *jc,
...
     if (strcmp(entry->key, custom_metric_value_type) == 0) {
-      metric_value_type = strdup(entry->value.value_text);
+      if (strcmp(entry->value.value_text, "INT64") == 0) {
+        metric_value_type = METRIC_VALUE_TYPE_INT64;
+      } else if (strcmp(entry->value.value_text, "DOUBLE") == 0) {
+        metric_value_type = METRIC_VALUE_TYPE_DOUBLE;
+      } else if (strcmp(entry->value.value_text, "BOOL") == 0) {
+        metric_value_type = METRIC_VALUE_TYPE_BOOL;
+      } else {
+        WARNING("write_gcm: unsupported metric value type provided: %s",
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c (hunk 11 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c (hunk 12 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done, called the helper function `wg_default_metric_value_type_from_ds_type` as I had access to the `ds_type` in all invocations.

File: src/write_gcm.c (hunk 13 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c (hunk 14 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c (hunk 6 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: This was an artifact from prior changes, I restored the original ordering.

File: src/write_gcm.c (hunk 15 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Ack via `UNKNOWN` equivalent.

File: src/write_gcm.c (hunk 16 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

File: src/write_gcm.c (hunk 17 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Done.

**Comment by bmoyles0117** [comment]
Everything should be good to go, PTAL.

File: src/docker.c [hunk 23]
```diff
@@ -33,7 +33,7 @@
 #include <curl/curl.h>
 
 // Default size of the response buffer when calling the Docker stats API
-#define RESPONSE_BUFFER_SIZE 128000
+#define RESPONSE_BUFFER_SIZE 1024000
```
Comment (igorpeshansky) [inline src/docker.c]: Unrelated to this PR, right? Can we pull this change out into a separate PR?

File: src/write_gcm.c [hunk 24]
```diff
@@ -1421,8 +1438,16 @@ static int wg_typed_value_create_from_value_t_inline(wg_typed_value_t *result,
...
-      result->field_name_static = "doubleValue";
-      snprintf(buffer, sizeof(buffer), "%f", value.gauge);
+      switch (metric_value_type) {
+        case METRIC_VALUE_TYPE_INT64:
+          result->field_name_static = "int64Value";
+          snprintf(buffer, sizeof(buffer), "%" PRIi64, (int64_t)value.gauge);
+          break;
+        case METRIC_VALUE_TYPE_DOUBLE:
+          result->field_name_static = "doubleValue";
+          snprintf(buffer, sizeof(buffer), "%f", value.gauge);
```
Comment (igorpeshansky) [inline src/write_gcm.c]: `double` requires `%lf`. Also below. Yes, I know it was like this in the original code, but we should fix it.

File: src/write_gcm.c [hunk 25]
```diff
@@ -3105,6 +3146,64 @@ static void wg_json_Metric(json_ctx_t *jc,
   wg_json_map_close(jc);
 }
 
+static int wg_default_metric_value_type_from_ds_type(int ds_type) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: How about s/`from`/`for`/, i.e., `wg_default_metric_value_type_for_ds_type`?

File: src/write_gcm.c [hunk 26]
```diff
@@ -3105,6 +3146,64 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+  int metric_value_type = METRIC_VALUE_TYPE_UNKNOWN;
+
+  switch(ds_type) {
+    case DS_TYPE_GAUGE:
+      metric_value_type = METRIC_VALUE_TYPE_DOUBLE;
```
Comment (igorpeshansky) [inline src/write_gcm.c]: You can just return it here... Then you don't need a local at all.

File: src/write_gcm.c [hunk 27]
```diff
@@ -3105,6 +3146,64 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+}
+
+static int wg_metric_value_type_from_wg_payload_t(
+    const wg_payload_t *element) {
+  int metric_value_type = METRIC_VALUE_TYPE_UNKNOWN;
+
+  for (int i = 0; i < element->key.num_metadata_entries; ++i) {
+    wg_metadata_entry_t *entry = &element->key.metadata_entries[i];
+    if (strcmp(entry->key, custom_metric_value_type_key) == 0) {
+      if (strcmp(entry->value.value_text, "INT64") == 0) {
```
Comment (igorpeshansky) [inline src/write_gcm.c]: This could be a loop over the `metric_value_type_strings` array, right?

File: src/write_gcm.c [hunk 28]
```diff
@@ -3105,6 +3146,64 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+        metric_value_type = METRIC_VALUE_TYPE_INT64;
+      } else if (strcmp(entry->value.value_text, "DOUBLE") == 0) {
+        metric_value_type = METRIC_VALUE_TYPE_DOUBLE;
+      } else if (strcmp(entry->value.value_text, "BOOL") == 0) {
+        metric_value_type = METRIC_VALUE_TYPE_BOOL;
+      } else {
+        WARNING("write_gcm: ignoring metric value type provided: %s, will use"
+                "default instead.", 
+                entry->value.value_text);
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Fits on the previous line.

File: src/write_gcm.c [hunk 29]
```diff
@@ -3105,6 +3146,64 @@ static void wg_json_Metric(json_ctx_t *jc,
...
+
+static int wg_metric_value_type_from_wg_payload_t(
+    const wg_payload_t *element) {
+  int metric_value_type = METRIC_VALUE_TYPE_UNKNOWN;
+
+  for (int i = 0; i < element->key.num_metadata_entries; ++i) {
+    wg_metadata_entry_t *entry = &element->key.metadata_entries[i];
+    if (strcmp(entry->key, custom_metric_value_type_key) == 0) {
+      if (strcmp(entry->value.value_text, "INT64") == 0) {
+        metric_value_type = METRIC_VALUE_TYPE_INT64;
```
Comment (igorpeshansky) [inline src/write_gcm.c]: If you return here, then you'll know that at the end of the outer loop the value is still unknown, making the code much simpler...

File: src/write_gcm.c [hunk 30]
```diff
@@ -3205,7 +3306,7 @@ static int wg_json_CreateTimeSeries(
             head->key.type_instance);
       continue;
     }
-    if (head->values[0].ds_type == DS_TYPE_GAUGE
+    if ((head->values[0].ds_type == DS_TYPE_GAUGE)
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Why did you need to introduce these parentheses?

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: src/write_gcm.c (hunk 24 above)
Comment (kosak) [reply src/write_gcm.c]: It is not correct to say that double "requires" ```%lf```. 

It's more correct to say that for [fsn]printf, ```%f``` is the correct specifier and ```%lf``` is permitted an an alternative... the "l" is a no-op... it has no effect on %e, %f, %g and friends).

(This is NOT the case for the scanf family, where the exact type is required).

The reason the printf family can be looser like this is that variadic arguments undergo promotion... so all chars become ints, floats become doubles, etc.

Comment (kosak) [reply src/write_gcm.c]: ...and the compiler would have caught it if it were actually a problem, because gcc does format checking these days.

File: src/docker.c (hunk 23 above)
Comment (bmoyles0117) [reply src/docker.c]: We can remove this, as @dhrupadb made a change in his branch to reflect the same thing. This was just committed so I could produce a build out of it.

File: src/write_gcm.c (hunk 24 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: Should we update this? WDYT?

**Comment by bmoyles0117** [comment]
I've made most of the updates as requested, PTAL.

File: src/write_gcm.c (hunk 27 above)
Comment (igorpeshansky) [reply src/write_gcm.c]: Not done.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

**Comment by bmoyles0117** [comment]
Anything stopping this from being approved at this point?

**Comment by igorpeshansky** [comment]
There's one remaining comment [here](https://github.com/Stackdriver/collectd/pull/103/files/8d51abdf1d239f5b65fa9ac0a091c363c84da590#diff-f54cd6830a1d91892682e12d794bde80R3172).

File: src/write_gcm.c (hunk 27 above)
Comment (bmoyles0117) [reply src/write_gcm.c]: I'm torn on this, I feel as though it might be too magical by using `metric_value_type_strings`, here's the code I would use in its place, and I'm not sure that the meaning is clear to the end user.

```
      for (int j = 0; j < STATIC_ARRAY_SIZE(metric_value_type_strings); j++) {
        if (strcmp(entry->value.value_text, metric_value_type_strings[j]) == 0) {
          return j;
        }
      }
      
      WARNING("write_gcm: ignoring metric value type provided: %s, will use"
              "default instead.", entry->value.value_text);
      break;
```

WDYT? Is there a better more intuitive way to use the array without losing meaning? If you feel that this is clear then I'll push a commit for it.

Comment (bmoyles0117) [reply src/write_gcm.c]: @igorpeshansky friendly ping for your thoughts here. Looking to validate whether the proposed solution is still "developer friendly", or if we should stay with the current implementation due to its explicitness.

**Comment by jkohen** [comment]
@bmoyles0117 what do we need in order to resolve this PR?

**Comment by bmoyles0117** [comment]
As this PR is almost 2 years old, I've lost most context. @igorpeshansky do you feel that this change is still relevant / necessary, or can we abandon?
